rustls/src/crypto/ring/quic.rs

```diff
@@ -0,0 +1,430 @@
+use crate::crypto::cipher::Iv;
+use crate::error::Error;
+use crate::hkdf;
+use crate::quic;
+use crate::tls13::key_schedule::{hkdf_expand, hkdf_expand_aead_key};
+use crate::tls13::Tls13CipherSuite;
+
+use ring::aead;
+
+pub(crate) struct HeaderProtectionKey(aead::quic::HeaderProtectionKey);
+
+impl HeaderProtectionKey {
+    pub(crate) fn new(
+        expander: &hkdf::Expander,
+        version: quic::Version,
+        alg: &'static aead::quic::Algorithm,
+    ) -> Self {
+        let key = hkdf_expand_aead_key(expander, alg.key_len(), version.header_key_label(), &[]);
+        Self(aead::quic::HeaderProtectionKey::new(alg, key.as_ref()).unwrap())
+    }
+
+    fn xor_in_place(
+        &self,
+        sample: &[u8],
+        first: &mut u8,
+        packet_number: &mut [u8],
+        masked: bool,
+    ) -> Result<(), Error> {
+        // This implements "Header Protection Application" almost verbatim.
+        // <https://datatracker.ietf.org/doc/html/rfc9001#section-5.4.1>
+
+        let mask = self
+            .0
+            .new_mask(sample)
+            .map_err(|_| Error::General("sample of invalid length".into()))?;
+
+        // The `unwrap()` will not panic because `new_mask` returns a
+        // non-empty result.
+        let (first_mask, pn_mask) = mask.split_first().unwrap();
+
+        // It is OK for the `mask` to be longer than `packet_number`,
+        // but a valid `packet_number` will never be longer than `mask`.
+        if packet_number.len() > pn_mask.len() {
+            return Err(Error::General("packet number too long".into()));
+        }
+
+        // Infallible from this point on. Before this point, `first` and
+        // `packet_number` are unchanged.
+
+        const LONG_HEADER_FORM: u8 = 0x80;
+        let bits = match *first & LONG_HEADER_FORM == LONG_HEADER_FORM {
+            true => 0x0f,  // Long header: 4 bits masked
+            false => 0x1f, // Short header: 5 bits masked
+        };
+
+        let first_plain = match masked {
+            // When unmasking, use the packet length bits after unmasking
+            true => *first ^ (first_mask & bits),
+            // When masking, use the packet length bits before masking
+            false => *first,
+        };
+        let pn_len = (first_plain & 0x03) as usize + 1;
+
+        *first ^= first_mask & bits;
+        for (dst, m) in packet_number
+            .iter_mut()
+            .zip(pn_mask)
+            .take(pn_len)
+        {
+            *dst ^= m;
+        }
+
+        Ok(())
+    }
+}
+
+impl quic::HeaderProtectionKey for HeaderProtectionKey {
+    fn encrypt_in_place(
+        &self,
+        sample: &[u8],
+        first: &mut u8,
+        packet_number: &mut [u8],
+    ) -> Result<(), Error> {
+        self.xor_in_place(sample, first, packet_number, false)
```

Comment:
this isn't specific to the crypto provider work, so feel free to punt, but it feels like a place where an `enum` could replace a `bool` to increase specificity and make it easier to distinguish what the call-sites intend without referencing the `xor_in_place` impl.

Reply:
i think i'll defer this to a later PR to avoid growing this one any further